modules/aws/src/smithy4s/query/AwsQueryCodecAPI.scala

```diff
@@ -16,47 +16,91 @@
 
 package smithy4s.aws.query
 
+import cats.effect.SyncIO
+import cats.syntax.either._
 import smithy4s.http.{BodyPartial, CodecAPI, HttpMediaType, PayloadError}
 import smithy4s.schema.CompilationCache
 import smithy4s.{PayloadPath, Schema}
+import smithy4s.xml.internals.XmlDecoder
+import smithy4s.xml.internals.XmlCursor
+import smithy4s.xml.internals.XmlDecoderSchemaVisitor
+import smithy4s.internals.InputOutput
+import smithy4s.xml.XmlDocument
+import fs2.Stream
+import fs2.data.xml._
+import fs2.data.xml.dom._
 
 import java.nio.ByteBuffer
 
 private[aws] class AwsQueryCodecAPI() extends CodecAPI {
-  private val decodingError = Left(
-    PayloadError(
-      PayloadPath.root,
-      "decoding is unsupported, encoding only",
-      "TODO: integrate with XML decoders"
-    )
-  )
-
-  override type Codec[A] = AwsQueryCodec[A]
-  override type Cache = CompilationCache[AwsQueryCodec]
 
-  override def createCache(): Cache = CompilationCache.make[Codec]
+  override type Codec[A] = Either[AwsQueryCodec[A], XmlDecoder[A]]
+  override type Cache = CompilationCache[AwsQueryCodec]
 
-  override def mediaType[A](codec: Codec[A]): HttpMediaType =
-    HttpMediaType("application/x-www-form-urlencoded")
+  override def createCache(): Cache = CompilationCache.make[AwsQueryCodec]
 
   override def compileCodec[A](
       schema: Schema[A],
       cache: Cache
-  ): Codec[A] = schema.compile(new AwsSchemaVisitorAwsQueryCodec(cache))
+  ): Codec[A] =
+    schema.hints match {
+      case InputOutput.hint(InputOutput.Input) =>
+        Left(schema.compile(new AwsSchemaVisitorAwsQueryCodec(cache)))
+      case InputOutput.hint(InputOutput.Output) =>
+        Right(schema.compile(XmlDecoderSchemaVisitor))
+    }
+
+  override def mediaType[A](codec: Codec[A]): HttpMediaType =
+    HttpMediaType("application/x-www-form-urlencoded")
 
   override def decodeFromByteArrayPartial[A](
       codec: Codec[A],
       bytes: Array[Byte]
-  ): Either[PayloadError, BodyPartial[A]] = decodingError
+  ): Either[PayloadError, BodyPartial[A]] = codec match {
+    case Left(_) =>
+      Left(
+        PayloadError(
+          PayloadPath.root,
+          "",
+          "Invalid codec: got AWS Query encoder, expected XML decoder"
+        )
+      )
+    case Right(xmlDecoder) =>
+      Stream
+        .emit[SyncIO, String](new String(bytes, "UTF-8"))
+        .through(events[SyncIO, String]())
+        .through(documents[SyncIO, XmlDocument])
+        .map(doc => XmlCursor.fromDocument(doc))
+        .map(cursor => xmlDecoder.decode(cursor))
+        .rethrow
+        .head
+        .compile
+        .lastOrError
+        .map(a =>
+          // TODO: have no idea if it's correct or not
+          BodyPartial(Map.empty.withDefaultValue(a))
```

Comment:
I don't understand this part. What kind of a `Map[String, A]` should be created here for an `a: A`?

Reply:
Right, so the `partial` stuff is there because in some protocols, only parts of the data come from the http body. The rest may come from elsewhere,  http headers, etc. So the `BodyPartial` aims at capturing the default value. 

Here, we're decoding the totality of the of the body from the payload, so you can return `BodyPartial(_ => a)`, which means that data that comes from elsewhere is not needed for the decoding here.

Reply:
Done